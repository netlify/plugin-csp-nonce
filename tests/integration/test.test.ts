import parseContentSecurityPolicy from "content-security-policy-parser";
import * as cheerio from "cheerio";
import { readFile } from "node:fs/promises";

import { afterAll, beforeAll, expect, it, describe, beforeEach } from "vitest";
import { serve } from "./helpers";


function isCheerioAPIShape(val) {
  const props = ['contains', 'extract', 'html', 'merge', 'parseHTML', 'root', 'text', 'xml', 'load', '_root', '_options', 'fn'];
  for (const prop of props) {
    if (!Object.hasOwn(val, prop)) {
      return false;
    }
  }
  return true;
}
expect.addSnapshotSerializer({
  serialize($, config, indentation, depth, refs, printer) {
    const elements = $('script[nonce]:not([nonce=""]),link[rel="preload"][as="script"][nonce]:not([nonce=""])');
    for (const element of elements) {
      element.attribs.nonce = "<placeholder_for_snapshop_test>"
    }
    return $.html()
  },
  test(val) {
    return isCheerioAPIShape(val)
  },
})

let baseURL: string;

let cleanup: () => Promise<void>;

beforeAll(async () => {
  [baseURL, cleanup] = await serve({
    waitForFunctions: ["__csp-nonce"],
  });
});

afterAll(async () => {
  await cleanup()
});

describe("GET /", function () {
  let response: Response;
  beforeEach(async () => {
    response = await fetch(new URL(`/`, baseURL));
  });

  it("__csp-nonce edge function was invoked", () => {
    expect(response.headers.get("x-debug-csp-nonce")).to.eql("invoked");
  });

  it("responds with a 200 status", () => {
    expect(response.status).to.eql(200);
  });

  it("responds with a content-security-policy header", () => {
    expect(response.headers.has("content-security-policy")).to.eql(true);
  });

  describe("content-security-policy header", () => {
    let csp: Map<string, string[]>;
    beforeEach(async () => {
      csp = parseContentSecurityPolicy(
        response.headers.get("content-security-policy") || ""
      );
    });

    it("has correct img-src directive", () => {
      expect(csp.get("img-src")).to.eql(["'self'", "blob:", "data:"]);
    });
    it("has correct script-src directive", () => {
      const script = csp.get("script-src");
      const nonce = /^'nonce-[-A-Za-z0-9+/]{32}'$/;
      expect(script.find((value) => nonce.test(value))).to.match(nonce);
      expect(script.includes("'strict-dynamic'")).to.eql(true);
      expect(script.includes("'unsafe-inline'")).to.eql(true);
      expect(script.includes("'unsafe-eval'")).to.eql(true);
      expect(script.includes("'self'")).to.eql(true);
      expect(script.includes("https:")).to.eql(true);
      expect(script.includes("http:")).to.eql(true);
      expect(
        script.includes("'sha256-/Cb4VxgL2aVP0MVDvbP0DgEOUv+MeNQmZX4yXHkn/c0='")
      ).to.eql(true);
    });
    it("has correct report-uri directive", () => {
      expect(csp.get("report-uri")).to.eql([
        "/.netlify/functions/__csp-violations",
      ]);
    });
  });

  describe("html nonces", () => {
    let $: cheerio.CheerioAPI;
    beforeEach(async () => {
      $ = cheerio.load(await response.text());
    });

    it("has set the nonce attribute on the script elements", () => {
      const csp = parseContentSecurityPolicy(
        response.headers.get("content-security-policy") || ""
      );
      const scriptsrc = csp.get("script-src");
      const nonce = scriptsrc
        ?.find((v) => v.startsWith("'nonce-"))
        ?.slice("'nonce-".length, -1)!;

      const scripts = $("script");
      for (const script of scripts) {
        expect(script.attribs.nonce).to.eql(nonce);
      }
    });

    it("has set the nonce attribute on the link preload script elements", () => {
      const csp = parseContentSecurityPolicy(
        response.headers.get("content-security-policy") || ""
      );
      const scriptsrc = csp.get("script-src");
      const nonce = scriptsrc
        ?.find((v) => v.startsWith("'nonce-"))
        ?.slice("'nonce-".length, -1)!;

      const elements = $("link[rel=\"preload\"][as=\"script\"]");
      for (const element of elements) {
        expect(element.attribs.nonce).to.eql(nonce);
      }
    });
  });

  it('html snapshot with filtered nonce', async () => {
    expect(cheerio.load(await response.text())).toMatchSnapshot()
  })
});

describe("POST /", function () {
  let response: Response;
  beforeAll(async () => {
    response = await fetch(new URL(`/`, baseURL), {
      method: "POST",
    });
  });

  it("__csp-nonce edge function was not invoked", () => {
    expect(response.headers.has("x-debug-csp-nonce")).to.eql(false);
  });

  it("responds with original content-security-policy header", () => {
    expect(response.headers.get("content-security-policy")).to.eql(
      "img-src 'self' blob: data:; script-src 'sha256-/Cb4VxgL2aVP0MVDvbP0DgEOUv+MeNQmZX4yXHkn/c0='"
    );
  });
});

describe("GET /main.css", function () {
  let response: Response;
  beforeAll(async () => {
    response = await fetch(new URL(`/main.css`, baseURL));
  });

  it("__csp-nonce edge function was not invoked", () => {
    expect(response.headers.has("x-debug-csp-nonce")).to.eql(false);
  });

  it("responds with a 200 status", () => {
    expect(response.status).to.eql(200);
  });

  it("responds without a content-security-policy header", () => {
    expect(response.headers.has("content-security-policy")).to.eql(false);
  });
});


describe("Origin response has non-html content-type", () => {
  let response: Response;
  beforeAll(async () => {
    response = await fetch(new URL(`/hello`, baseURL));
  });

  it("__csp-nonce edge function was invoked", () => {
    expect(response.headers.get("x-debug-csp-nonce")).to.eql("invoked");
  });

  it("responds with a 200 status", () => {
    expect(response.status).to.eql(200);
  });

  it("responds without a content-security-policy header", () => {
    expect(response.headers.has("content-security-policy")).to.eql(false);
  });

  describe("body", () => {
    it("has has the original response body unmodified", async () => {
      const actual = Buffer.from(await response.arrayBuffer())
      const expected = await readFile(new URL("../../site/hello", import.meta.url))
      expect(actual).to.eql(expected)
    });
  });
})

describe("Origin response has html content-type but binary contents in body", () => {
  let response: Response;
  beforeAll(async () => {
    response = await fetch(new URL(`/i-am-really-a-png-file.html`, baseURL));
  });

  it("__csp-nonce edge function was invoked", () => {
    expect(response.headers.get("x-debug-csp-nonce")).to.eql("invoked");
  });

  it("responds with a 200 status", () => {
    expect(response.status).to.eql(200);
  });

  it("responds with a content-security-policy header", () => {
    expect(response.headers.has("content-security-policy")).to.eql(true);
  });

  describe("content-security-policy header", () => {
    let csp: Map<string, string[]>;
    beforeAll(async () => {
      csp = parseContentSecurityPolicy(
        response.headers.get("content-security-policy") || ""
      );
    });

    it("has correct img-src directive", () => {
      expect(csp.get("img-src")).to.eql(["'self'", "blob:", "data:"]);
    });
    it("has correct script-src directive and has not overridden the original nonce value", () => {
      const script = csp.get("script-src")!;
      const nonce = /^'nonce-[-A-Za-z0-9+/]{32}'$/;
      expect(script.find((value) => nonce.test(value))).to.match(nonce);
      expect(script.includes("'strict-dynamic'")).to.eql(true);
      expect(script.includes("'unsafe-inline'")).to.eql(true);
      expect(script.includes("'unsafe-eval'")).to.eql(true);
      expect(script.includes("'self'")).to.eql(true);
      expect(script.includes("https:")).to.eql(true);
      expect(script.includes("http:")).to.eql(true);
      expect(
        script.includes("'sha256-/Cb4VxgL2aVP0MVDvbP0DgEOUv+MeNQmZX4yXHkn/c0='")
      ).to.eql(true);
    });
    it("has correct report-uri directive", () => {
      expect(csp.get("report-uri")).to.eql([
        "/.netlify/functions/__csp-violations",
      ]);
    });
  });

  describe("body", () => {
    it("has has the original response body unmodified", async () => {
      const actual = Buffer.from(await response.arrayBuffer())
      const expected = await readFile(new URL("../../site/i-am-really-a-png-file.html", import.meta.url))
      expect(actual).to.eql(expected)
    });
  });
})

describe("Origin response has html content-type but non-html text contents in body", () => {
  let response: Response;
  beforeAll(async () => {
    response = await fetch(new URL(`/i-am-really-a-json-file.html`, baseURL));
  });

  it("__csp-nonce edge function was invoked", () => {
    expect(response.headers.get("x-debug-csp-nonce")).to.eql("invoked");
  });

  it("responds with a 200 status", () => {
    expect(response.status).to.eql(200);
  });

  it("responds with a content-security-policy header", () => {
    expect(response.headers.has("content-security-policy")).to.eql(true);
  });

  describe("content-security-policy header", () => {
    let csp: Map<string, string[]>;
    beforeAll(async () => {
      csp = parseContentSecurityPolicy(
        response.headers.get("content-security-policy") || ""
      );
    });

    it("has correct img-src directive", () => {
      expect(csp.get("img-src")).to.eql(["'self'", "blob:", "data:"]);
    });
    it("has correct script-src directive and has not overridden the original nonce value", () => {
      const script = csp.get("script-src")!;
      const nonce = /^'nonce-[-A-Za-z0-9+/]{32}'$/;
      expect(script.find((value) => nonce.test(value))).to.match(nonce);
      expect(script.includes("'strict-dynamic'")).to.eql(true);
      expect(script.includes("'unsafe-inline'")).to.eql(true);
      expect(script.includes("'unsafe-eval'")).to.eql(true);
      expect(script.includes("'self'")).to.eql(true);
      expect(script.includes("https:")).to.eql(true);
      expect(script.includes("http:")).to.eql(true);
      expect(
        script.includes("'sha256-/Cb4VxgL2aVP0MVDvbP0DgEOUv+MeNQmZX4yXHkn/c0='")
      ).to.eql(true);
    });
    it("has correct report-uri directive", () => {
      expect(csp.get("report-uri")).to.eql([
        "/.netlify/functions/__csp-violations",
      ]);
    });
  });

  describe("body", () => {
    it("has has the original response body unmodified", async () => {
      const actual = Buffer.from(await response.arrayBuffer())
      const expected = await readFile(new URL("../../site/i-am-really-a-json-file.html", import.meta.url))
      expect(actual).to.eql(expected)
    });
  });
})
