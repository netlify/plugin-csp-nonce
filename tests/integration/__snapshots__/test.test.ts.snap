// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET / > html snapshot with filtered nonce 1`] = `
<!DOCTYPE html><html><head>
    <title>Using CSP Nonces on Script Tags</title>
    <link rel="stylesheet" href="/main.css">
    <link rel="preload" as="script" href="https://cdn.jsdelivr.net/npm/jquery@3.7.0/dist/jquery.min.js" nonce="<placeholder_for_snapshop_test>">
  </head>
  <body>
    <h1>Using CSP Nonces on Script Tags</h1>
    <p>This is a simple demonstration of using nonces in script tags.</p>
    <a href="#" onclick="alert('ohnoes')">
      Clicking this should not trigger an alert
    </a>
    <script nonce="<placeholder_for_snapshop_test>">
      document.write("<p>✅ Inline script without HTML attributes</p>");
    </script>
    <script class="testing" data-whatever="sup" nonce="<placeholder_for_snapshop_test>">
      document.write("<p>✅ Inline script with HTML attributes</p>");
    </script>
    <script src="https://cdn.jsdelivr.net/npm/jquery@3.7.0/dist/jquery.min.js" nonce="<placeholder_for_snapshop_test>"></script>
    <script nonce="<placeholder_for_snapshop_test>">
      // check that jquery was downloaded from CDN
      if ($) {
        document.write("<p>✅ Script tag with external src</p>");
      }
    </script>
    <script nonce="<placeholder_for_snapshop_test>">
      const el = document.createElement("script");
      el.innerText =
        "document.write('<p>✅ Script tag from <code>document.createElement(\\"script\\")</code></p>');";
      document.body.appendChild(el);
    </script>
    <script nonce="<placeholder_for_snapshop_test>">
      // if the nonce works, then this shouldn't run!
      document.write(
        "<scr" +
          "ipt>" +
          "document.write('<div>❌ Script tag from <code>document.write</code> has executed!</div>');" +
          "</scr" +
          "ipt>"
      );
    </script>
    <script nonce="<placeholder_for_snapshop_test>">
      // test for https://github.com/netlify/plugin-csp-nonce/issues/72 case
      document.write(
        "<scr" +
          "ipt>" +
          "document.write('<p>✅ User-defined script-src value is preserved!</p>');" +
          "</scr" +
          "ipt>"
      );
    </script>
    <script nonce="<placeholder_for_snapshop_test>">
      var CHECKS_COUNT = 6;
      if (document.getElementsByTagName("p").length === CHECKS_COUNT) {
        document.write("<p>✅ Test suite has succeeded</p>");
      } else {
        document.write("<div>❌ Test suite has failed</div>");
      }
    </script>
  

</body></html>
`;
